### ISSUE TITLE
Gateway Timeout on Docker Swarm worker replicas

### ISSUE THREAD

**Issue by statickidz** [Dokploy/dokploy#592]
### To Reproduce

1. Create a Dokploy simply Docker Swarm configuration with 1 manager and 1 worker.

    ![image](https://github.com/user-attachments/assets/73dfd655-e6e4-4381-b41c-0e54e847c5c4)

3. Create an app with https://github.com/Dokploy/swarm-test

    ![image](https://github.com/user-attachments/assets/f93e0335-3b77-4e8c-993b-5745365be403)

4. Put more than 1 replica in the Swarm config

    ![image](https://github.com/user-attachments/assets/f8b0ba0a-4903-489a-84f5-fb3a62e63343)

5. Verify all deployed replicas are splitting well in the two instances

    **Manager**
    ![image](https://github.com/user-attachments/assets/edace26f-0d68-444e-804b-cc4f40f42112)

    **Worker**
    ![image](https://github.com/user-attachments/assets/94045005-ceb0-4f91-a853-ae9627d7d5f1)

### Current vs. Expected behavior

I expect all the Docker Swarm containers work normally independently where the request goes both on manager and worker instances but it seems like when the request goes to worker instance I get Gateway Timeout, otherwise if it goes to manager works.

- Example 1: Request > Swarm decides to go manager > Works
- Example 2: Request > Swarm decides to go worker > Gateway Timeout

### Provide environment information

```bash
Operating System:
  OS: Canonical-Ubuntu-22.04-aarch64-2024.06.26-0
  Arch : arm64
Dokploy version: v0.10.3
VPS Provider: Oracle Cloud
What application/services are you trying to deploy?: Simple Nodejs app
```

### Which area(s) are affected? (Select all that apply)

Application, Docker Compose, Traefik, Docker

### Additional context

To check that it's not a network issue between instances or something I created a rule to open all the ports in the security list, by the way I'm using this project to boot the instances: https://github.com/statickidz/dokploy-oci-free/

**Comment by Siumauricio** [comment]
hmm I think something is needed at the traefik level to make it able to route to the worker container.

**Comment by statickidz** [comment]
> hmm I think something is needed at the traefik level to make it able to route to the worker container.

Is this something related to my environment or you were able to make it work before?

**Comment by Siumauricio** [comment]
@statickidz Yes, this already worked for me some time ago, however since we upgraded traefik to version 3 I haven't tried it, surely there was some change.

**Comment by Siumauricio** [comment]
I recently tested and is working for me used this docker image


![Screenshot 2024-10-29 at 11 24 18 PM](https://github.com/user-attachments/assets/57f56074-d453-42fe-b0c1-9a4cde0bf3f3)

![Screenshot 2024-10-29 at 11 24 33 PM](https://github.com/user-attachments/assets/5d840221-368a-42b2-9fad-1dac6f5f6211)

![Screenshot 2024-10-29 at 11 24 39 PM](https://github.com/user-attachments/assets/3ace58c0-c060-4704-8725-97e58a8bc345)

I don't have any running container in the dokploy server

![image](https://github.com/user-attachments/assets/e6d25144-648e-4231-890d-06942a691536)

In the worker is running 6 instances
![Screenshot 2024-10-29 at 11 25 27 PM](https://github.com/user-attachments/assets/f3b3b8e1-dbfb-4c40-b475-84180550bd10)

The domain I've used
![Screenshot 2024-10-29 at 11 26 02 PM](https://github.com/user-attachments/assets/35773f53-0797-4482-9f3b-2306a2ce8eda)

and when you enter you will see this
![Screenshot 2024-10-29 at 11 26 22 PM](https://github.com/user-attachments/assets/bb14a0aa-df8c-4f38-b9f6-6dc24f1ee4a2)

If you reload after a couple minutes the information should change since is using another private ip and everything, so the load balancing working fine 
![Screenshot 2024-10-29 at 11 26 30 PM](https://github.com/user-attachments/assets/b8025aba-ffef-4112-b240-e8c47ff406a6)

**Comment by statickidz** [comment]
@Siumauricio I see! I just created a new Dokploy instances (manager and worker) in AWS to check if it was something related with OCI but I'm getting the same result, that's quite weird. As before, all ports opened, no issues joining the Swarm cluster but when the request leads to the worker I get the Gateway Timeout. At this point I'm not sure what could be.

**Comment by Siumauricio** [comment]
Did you make a custom installation? or did you installed with the official script?

**Comment by statickidz** [comment]
> Did you make a custom installation? or did you installed with the official script?

For the main instance official script, for the workers the commands provided on the "Add Node" button.

**Comment by Siumauricio** [comment]
I see you are exiting docker swarm in the worker, then how did you link the worker to the manager, you follow the steps from the Add Node button manually?

I would recommend you first try using the traditional way that dokploy gives, that is linking the workers manually, if you see that it works, I think it would be a problem of your infrastructure setup.

**Comment by binaryYuki** [comment]
Is your infrastructure running on Oracle? I encountered the same problem, but it runs normally if executed on the same node where Traefik is located.

**Comment by statickidz** [comment]
> Have you check in the dashboard of dokploy if you have the worker associated in the cluster section?

Yep, it's been displayed correctly

![image](https://github.com/user-attachments/assets/705df1db-81c1-4d84-89fc-9500c704a2ef)


> I see you are exiting docker swarm in the worker, then how did you link the worker to the manager, you follow the steps from the Add Node button manually?
> 
> I would recommend you first try using the traditional way that dokploy gives, that is linking the workers manually, if you see that it works, I think it would be a problem of your infrastructure setup.

Same result either if I pre-install docker and I pre-leave swarm (like in the script) or if I take the Dokploy quick steps to install it.

For example, this is the last test on a fresh worker node with the dokploy steps, result is always Gateway Timeout:

![image](https://github.com/user-attachments/assets/0ff962b5-e536-46c5-b50c-3d26c78c7e67)

@Siumauricio this is a test environment so if you feel you want to debug that in deep reach me, I can provide you the access to the instances

> Is your infrastructure running on Oracle OCI? I encountered the same problem, but it runs normally if executed on the same node where Traefik is located.

Found it on the Oracle OCI, works well if I point all the instances to the manager with this:

![image](https://github.com/user-attachments/assets/e38104e6-7949-44b5-9872-0d32790f37de)

But I feel this is not OCI related, because I created a couple of instances on AWS to try and the result was the same 
 https://github.com/Dokploy/dokploy/issues/592#issuecomment-2447020784

**Comment by binaryYuki** [comment]
> But I feel this is not OCI related, because I created a couple of instances on AWS to try and the result was the same

I just try it on my azure server and the same issue occurd.

@Siumauricio Can we try load balance of traefik like
```toml
[tcp.services]
  [tcp.services.app]
    [[tcp.services.app.weighted.services]]
      name = "appv1"
      weight = 3
    [[tcp.services.app.weighted.services]]
      name = "appv2"
      weight = 1

  [tcp.services.appv1]
    [tcp.services.appv1.loadBalancer]
      [[tcp.services.appv1.loadBalancer.servers]]
        address = "private-ip-server-1/:8080"

  [tcp.services.appv2]
    [tcp.services.appv2.loadBalancer]
      [[tcp.services.appv2.loadBalancer.servers]]
        address = "private-ip-server-2/:8080"
```
instead of pointing them directly to the service itself like
```yaml
  services:
    animeapi-core-409c00-service-11:
      loadBalancer:
        servers:
          - url: http://animeapi-core-409c00:8000
```

**Comment by aliuq** [comment]
> I recently tested and is working for me used this docker image我最近进行了测试，并为我工作使用了此Docker图像

Visit `xxx.traefik.me` got error `Bad Gateway` with version `v0.17.9`

![Image](https://github.com/user-attachments/assets/81f184d8-ff19-4cf4-81c0-2795b89fa73e)

![Image](https://github.com/user-attachments/assets/7ba7089e-d818-4fa1-a75a-6a6fc6d2c92f)

![Image](https://github.com/user-attachments/assets/0db3990a-c553-4e14-b575-de7367658350)

**Comment by statickidz** [comment]
> > I recently tested and is working for me used this docker image我最近进行了测试，并为我工作使用了此Docker图像
> 
> Visit `xxx.traefik.me` got error `Bad Gateway` with version `v0.17.9`
> 
> ![Image](https://github.com/user-attachments/assets/81f184d8-ff19-4cf4-81c0-2795b89fa73e)
> 
> ![Image](https://github.com/user-attachments/assets/7ba7089e-d818-4fa1-a75a-6a6fc6d2c92f)
> 
> ![Image](https://github.com/user-attachments/assets/0db3990a-c553-4e14-b575-de7367658350)

which cloud are you using to run your instances?

**Comment by codeit-ninja** [comment]
I get the same error. Swarm is not working, every service I deploy results into a `bad gateway` error.

In this example I used the same service as @Siumauricio 

![Image](https://github.com/user-attachments/assets/1d63bfdb-e41a-43e3-873e-9607ef823045)

![Image](https://github.com/user-attachments/assets/afb1c125-9112-496e-95f2-0122e285eeb3)

![Image](https://github.com/user-attachments/assets/bf4d06a7-d235-4ed6-bd93-a3b63e3231af)

Swarm is pretty useless this way, I have tried to debug but cant figure it out.

**Comment by sundakai** [comment]
Oracle Cloud has the same error that has been bugging me for a couple days now.
Access on manager deployed with compose works fine.
When deployed with Stack on the worker, it reports multiple errors
![Image](https://github.com/user-attachments/assets/a534ed1f-4433-413d-bc7e-1b85d74d76ed)
![Image](https://github.com/user-attachments/assets/81de64ff-e295-4967-959c-3d4f15833246)

**Comment by agustints** [comment]
Same issue on my end, is there any update on this?

**Comment by Hoshino-Yumetsuki** [comment]
> 我收到相同的错误。Swarm 无法正常工作，我部署的每个服务都会导致错误。`bad gateway`
> 
> 在此示例中，我使用了与
> 
> ![Image](https://github.com/user-attachments/assets/1d63bfdb-e41a-43e3-873e-9607ef823045)
> 
> ![Image](https://github.com/user-attachments/assets/afb1c125-9112-496e-95f2-0122e285eeb3)
> 
> ![Image](https://github.com/user-attachments/assets/bf4d06a7-d235-4ed6-bd93-a3b63e3231af)
> 
> Swarm 这种方式很没用，我试过调试但想不通。我的猜测是网络以某种方式搞砸了，我很确定 traefik 无法连接到 worker 容器。但是不知道怎么调试那个..
> 
> 我添加到 treafik.yml 这是 traefik 中引发的错误`level: DUBUG`
> 
> ```
> 2025-03-13T18:22:30Z DBG github.com/traefik/traefik/v3/pkg/server/service/proxy.go:100 > 502 Bad Gateway error="dial tcp: lookup test-whoami-tegwkb on 127.0.0.11:53: no such host"
> ```

I'm also having the same issue

**Comment by sundakai** [comment]
> Maybe we should run a traefik on each node in the swarm, and then reverse proxy to the traefik on the node through the entrypoint, and then reverse proxy to the instance through the traefik in the node

No, the cluster should only have one web portal, check out this scenario of mine [manually install](https://github.com/Dokploy/dokploy/issues/592#issuecomment-2742074713)，it's work now.

![Image](https://github.com/user-attachments/assets/120cacf8-fcac-4f64-a551-ec0116ab5668)
![Image](https://github.com/user-attachments/assets/f3a7760c-33d1-472d-bb0f-ed7a05a7b7ad)
![Image](https://github.com/user-attachments/assets/a1b4b5aa-9fac-477a-a53f-3d8eaaea4cf6)

To a certain extent, dokploy is still not that stable, but I really like its design, similar to a cloud service.Some features are still missing, such as swarm management where you can only see containers on the master node. So I also installed portainer

**Comment by Hoshino-Yumetsuki** [comment]
> > Maybe we should run a traefik on each node in the swarm, and then reverse proxy to the traefik on the node through the entrypoint, and then reverse proxy to the instance through the traefik in the node
> 
> No, the cluster should only have one web portal, check out this scenario of mine [manually install](https://github.com/Dokploy/dokploy/issues/592#issuecomment-2742074713)，it's work now.
> 
> ![Image](https://github.com/user-attachments/assets/120cacf8-fcac-4f64-a551-ec0116ab5668) ![Image](https://github.com/user-attachments/assets/f3a7760c-33d1-472d-bb0f-ed7a05a7b7ad) ![Image](https://github.com/user-attachments/assets/a1b4b5aa-9fac-477a-a53f-3d8eaaea4cf6)
> 
> To a certain extent, dokploy is still not that stable, but I really like its design, similar to a cloud service.Some features are still missing, such as swarm management where you can only see containers on the master node. So I also installed portainer

Can I restore Dokploy from a backup after a manual installation? From what I have observed, restoring an instance from a backup will reset the docker network

**Comment by ron-tayler** [comment]
I have encountered a similar problem, but I have experience solving it. It is enough to add a "Label" so that Traeffik correctly searches for the IP of the container.

```
    deploy:
      labels:
        - traefik.docker.network=dokploy-network
```

**Comment by Hoshino-Yumetsuki** [comment]
This issue doesn't seem to be resolved, and I'm still getting the Gateway timeout error in my updated deployment

**Comment by ron-tayler** [comment]
> This issue doesn't seem to be resolved and I'm still getting a gateway timeout error in my deployment after the Dokploy update

I updated Dokploy and it solved the problem with finding a network for traefik. Maybe i can help you?
Pls send your swarm compose.

**Comment by ron-tayler** [comment]
I've considered a few more scenarios for the problem. I think this may be due to the fact that Traefik does not use the overlay Network. I'll try to check how it works with Dokploy. But I think that's where the problem lies.

**Comment by Hoshino-Yumetsuki** [comment]
> I had the same problem a couple of weeks ago. My application sporadically received a 502 bad gateway error. From what I could determine, the request never reached the container; it remained in Traefik. I spent several days investigating and applying Traefik configurations but couldn't solve it. In the end, I deleted the Traefik container and replaced it with Nginx Proxy Manager, and the problem never returned.
> 
> Dokploy can run without Traefik.

It's a good idea, but now that Dokploy is deeply integrated with Traefik, it might be difficult to migrate to a different solution

**Comment by Siumauricio** [comment]
Currently traefik is only a single part of dokploy, there's a feature request for caddy, I can consider work on support caddy which I consider it shouldn't be to much work

**Comment by sundakai** [comment]
> Currently traefik is only a single part of dokploy, there's a feature request for caddy, I can consider work on support caddy which I consider it shouldn't be to much work

I like Traefik better to a certain extent.
reverse proxies and control routes by means of labels really cool.
I installed docker swarm and dokploy manually , the problem never happened again.
 there must be a hidden bug! I understand It's not easy to fix.

**Comment by Siumauricio** [comment]
Mind to share the steps how did you installed manually? Maybe there's a bug in the instalation script

**Comment by Siumauricio** [comment]
Someone on Discord told me we should use a private IP instead of public Ip     https://github.com/Dokploy/website/blob/f886a19b446f2cf8686b7309318f13178699afb5/apps/website/public/install.sh#L87

And he told me it worked but I havent had time to test it

**Comment by sundakai** [comment]
> 介意分享您是如何手动安装的步骤的吗？也许 ininstallation 脚本中有 bug

sure, I've replied to this before. [manually install](https://github.com/Dokploy/dokploy/issues/592#issuecomment-2742074713), The docker swarm node does use an intranet.

**Comment by sundakai** [comment]
> Mind to share the steps how did you installed manually? Maybe there's a bug in the instalation script

sure ,I've replied to this before. [manually install](https://github.com/Dokploy/dokploy/issues/592#issuecomment-2742074713)
The docker swarm node does use an intranet.

**Comment by sundakai** [comment]
> Someone on Discord told me we should use a private IP instead of public Ip https://github.com/Dokploy/website/blob/f886a19b446f2cf8686b7309318f13178699afb5/apps/website/public/install.sh#L87
> 
> And he told me it worked but I havent had time to test it
> 
> I feel is something with docker swarm itself, like how we configure it in the instalation script

Whether multiple nodes are on the same intranet, this may not be as easy to determine, in some cases two nodes are indeed not on the same intranet, but When I manually installed,The docker swarm node does use a private IP.

**Comment by quantumlink** [comment]
I didn't have to remove dokploy first or recreate. 
Having managers and workers leave and then rejoining under the private IP fixed issues for me. So +1 to that theory and its evidence.

**Comment by Siumauricio** [comment]
I updated the script in the installation script to use private Ip instead of public Ip https://github.com/Dokploy/website/commit/b46705a57221dbcdd147575108050021fe9c5b13#diff-c659113fbc45a39fb618f85acb34da844d5a927faec06a4f77dc29cab9d0d6b6R87, I suggest to run `docker swarm leave --force` and then the installation script `curl -sSL https://dokploy.com/install.sh | sh` for new installations you don't need to do anything

**Comment by statickidz** [comment]
> I updated the script in the installation script to use private Ip instead of public Ip [Dokploy/website@b46705a#diff-c659113fbc45a39fb618f85acb34da844d5a927faec06a4f77dc29cab9d0d6b6R87](https://github.com/Dokploy/website/commit/b46705a57221dbcdd147575108050021fe9c5b13#diff-c659113fbc45a39fb618f85acb34da844d5a927faec06a4f77dc29cab9d0d6b6R87), I suggest to run `docker swarm leave --force` and then the installation script `curl -sSL https://dokploy.com/install.sh | sh` for new installations you don't need to do anything

It's working for me now on a new installation!

**Comment by Siumauricio** [comment]
Nice!! @statickidz Let's close this issue for now if you still are having this issue feel free to reopen
